src/main/kotlin/uk/gov/justice/digital/hmpps/welcometoprison/model/MovementService.kt

```diff
@@ -1,12 +1,59 @@
 package uk.gov.justice.digital.hmpps.welcometoprison.model
 
+import org.slf4j.LoggerFactory
 import org.springframework.stereotype.Service
+import uk.gov.justice.digital.hmpps.welcometoprison.model.MoveType.PRISON_TRANSFER
 import uk.gov.justice.digital.hmpps.welcometoprison.model.basm.BasmService
 import uk.gov.justice.digital.hmpps.welcometoprison.model.prison.PrisonService
+import uk.gov.justice.digital.hmpps.welcometoprison.model.prisonersearch.PrisonerSearchService
+import uk.gov.justice.digital.hmpps.welcometoprison.model.prisonersearch.response.MatchPrisonerResponse
 import java.time.LocalDate
 
 @Service
-class MovementService(val basmService: BasmService, val prisonService: PrisonService) {
+class MovementService(
+  val basmService: BasmService,
+  val prisonService: PrisonService,
+  val prisonerSearchService: PrisonerSearchService
+) {
   fun getMovements(agencyId: String, date: LocalDate) =
     basmService.getMoves(agencyId, date, date) + prisonService.getMoves(agencyId, date)
+
+  fun getMovementsMatchedWithPrisoner(agencyId: String, date: LocalDate): List<Movement> {
+    val movements = getMovements(agencyId, date)
+    return getMovementsMatchedWithPrisoner(movements)
+  }
+
+  private fun getMovementsMatchedWithPrisoner(movements: List<Movement>): List<Movement> {
+    fun findPrisonerMatch(identifier: String?, identifierName: String): MatchPrisonerResponse? {
+      val matchesByPrisonNumber = identifier?.let { prisonerSearchService.matchPrisoner(it) } ?: emptyList()
+      if (matchesByPrisonNumber.size > 1) log.warn("Multiple matched Prison records for a Movement by $identifierName. There are ${matchesByPrisonNumber.size} matched Prison records for $identifier")
+      return matchesByPrisonNumber.firstOrNull()
+    }
+    fun decorateMovementWithPrisonerMatch(movement: Movement): Movement {
+      var prisonNumberToUse = movement.prisonNumber
+      var pncToUse = movement.pncNumber
+
+      val matchByPrisonNumber = findPrisonerMatch(movement.prisonNumber, "Prison Number")
+      val matchByPncNumber = findPrisonerMatch(movement.pncNumber, "PNC Number")
+
+      if (matchByPrisonNumber === null && matchByPncNumber === null) prisonNumberToUse = null
+      if (matchByPrisonNumber === null && matchByPncNumber !== null) prisonNumberToUse = matchByPncNumber.prisonerNumber
+      if (matchByPrisonNumber !== null && matchByPncNumber === null) pncToUse = matchByPrisonNumber.pncNumber
+
+      return movement.copy(
+        prisonNumber = prisonNumberToUse,
+        pncNumber = pncToUse
+      )
+    }
+
+    var movementsMatchedWithPrisoner: List<Movement> = emptyList()
```

Comment:
As you're just transforming each value in a collection, this can be modelled as a `map`.
This will also get rid of the mutation of the reference `var` -> `val`

BASM won't return prison transfers going forward (DCS-1201) so safe to just wrap the BASM moves with this method.

Reply:
Removed the check for prison transfers, and changed it to a mutable list temporarily for the loop (which got rid of the var). Not too sure how the Map would help in this situation? Been a while since I've needed to use the Java Collections API though 😂